app/keepers/keepers.go

```diff
@@ -76,7 +134,442 @@ type AppKeepers struct {
+	)
+}
+
+// InitSpecialKeepers initiates special keepers (crisis appkeeper, upgradekeeper, params keeper)
+func (a *AppKeepers) InitSpecialKeepers(
+	appCodec codec.Codec,
+	bApp *baseapp.BaseApp,
+	cdc *codec.LegacyAmino,
+	invCheckPeriod uint,
+	skipUpgradeHeights map[int64]bool,
+	homePath string,
+) {
+	a.ParamsKeeper = initParamsKeeper(appCodec, cdc, a.keys[paramstypes.StoreKey], a.tkeys[paramstypes.TStoreKey])
+	// set the BaseApp's parameter store
+	bApp.SetParamStore(a.ParamsKeeper.Subspace(baseapp.Paramspace).WithKeyTable(paramstypes.ConsensusParamsKeyTable()))
+
+	// add capability keeper and ScopeToModule for ibc module
+	a.CapabilityKeeper = capabilitykeeper.NewKeeper(appCodec, a.keys[capabilitytypes.StoreKey], a.memKeys[capabilitytypes.MemStoreKey])
+
+	// grant capabilities for the ibc and ibc-transfer modules
+	a.ScopedIBCKeeper = a.CapabilityKeeper.ScopeToModule(ibchost.ModuleName)
+	a.ScopedTransferKeeper = a.CapabilityKeeper.ScopeToModule(ibctransfertypes.ModuleName)
+
+	a.CapabilityKeeper.Seal()
+
+	a.CrisisKeeper = crisiskeeper.NewKeeper(
+		a.GetSubspace(crisistypes.ModuleName), invCheckPeriod, a.BankKeeper, authtypes.FeeCollectorName,
+	)
+
+	a.UpgradeKeeper = upgradekeeper.NewKeeper(
+		skipUpgradeHeights,
+		a.keys[upgradetypes.StoreKey],
+		appCodec,
+		homePath,
+		bApp,
+		authtypes.NewModuleAddress(govtypes.ModuleName).String(),
+	)
+}
+
+func (a *AppKeepers) InitTransferStack() {
+	a.TransferStack = ibctransfer.NewIBCModule(a.TransferKeeper)
+	a.TransferStack = bridgingfee.NewIBCModule(
+		a.TransferStack.(ibctransfer.IBCModule),
+		a.DelayedAckKeeper,
+		a.TransferKeeper,
+		a.AccountKeeper.GetModuleAddress(txfeestypes.ModuleName),
+		a.RollappKeeper,
+	)
+	a.TransferStack = packetforwardmiddleware.NewIBCMiddleware(
+		a.TransferStack,
+		a.PacketForwardMiddlewareKeeper,
+		0,
+		packetforwardkeeper.DefaultForwardTransferPacketTimeoutTimestamp,
+		packetforwardkeeper.DefaultRefundTransferPacketTimeoutTimestamp,
+	)
+
+	a.TransferStack = denommetadatamodule.NewIBCModule(a.TransferStack, a.DenomMetadataKeeper, a.RollappKeeper)
+	// already instantiated in SetupHooks()
+	a.delayedAckMiddleware.Setup(
+		delayedackmodule.WithIBCModule(a.TransferStack),
+		delayedackmodule.WithKeeper(a.DelayedAckKeeper),
+		delayedackmodule.WithRollappKeeper(a.RollappKeeper),
+	)
+	a.TransferStack = a.delayedAckMiddleware
+	a.TransferStack = transfergenesis.NewIBCModule(a.TransferStack, a.DelayedAckKeeper, a.RollappKeeper, a.TransferKeeper, a.DenomMetadataKeeper)
+	a.TransferStack = transfergenesis.NewIBCModuleCanonicalChannelHack(a.TransferStack, a.RollappKeeper, a.IBCKeeper.ChannelKeeper)
+
+	// Create static IBC router, add transfer route, then set and seal it
+	ibcRouter := ibcporttypes.NewRouter()
+	ibcRouter.AddRoute(ibctransfertypes.ModuleName, a.TransferStack)
+	a.IBCKeeper.SetRouter(ibcRouter)
+}
+
+func (a *AppKeepers) SetupHooks() {
+	// register the staking hooks
+	a.StakingKeeper = *a.StakingKeeper.SetHooks(
```

Comment:
don't u overwrite here the `a.StakingKeeper` while already pass `&a.StakingKeeper` to different `.NewKeeper()` methods?